api/envoy/api/v2/route/route_components.proto

```diff
@@ -765,6 +786,33 @@ message RouteAction {
   //   requests to */prefix/etc* will be stripped to */etc*.
   string prefix_rewrite = 5;
 
+  // Indicates that during forwarding, portions of the path that match the
+  // pattern should be rewritten, even allowing the substitution of capture
+  // groups from the pattern into the new path as specified by the rewrite
+  // substitution string. This is useful to allow application paths to be
+  // rewritten in a way that is aware of segments with variable content like
+  // identifiers. The router filter will place the original path as it was
+  // before the rewrite into the :ref:`x-envoy-original-path
+  // <config_http_filters_router_x-envoy-original-path>` header.
+  //
+  // Examples:
+  //
+  // * The path pattern ``^/service/([^/]+)(/.*)$`` paired with a substitution
+  //   string of ``\2/instance/\1`` would transform ``/service/foo/v1/api``
+  //   into ``/v1/api/instance/foo``.
+  //
+  // * The pattern ``one`` paired with a substitution string of ``two`` would
+  //   transform ``/xxx/one/yyy/one/zzz`` into ``/xxx/two/yyy/two/zzz``.
+  //
+  // * The pattern ``^(.*?)one(.*)$`` paired with a substitution string of
+  //   ``\1two\2`` would replace only the first occurrence of ``one``,
+  //   transforming path ``/xxx/one/yyy/one/zzz`` into ``/xxx/two/yyy/one/zzz``.
+  //
+  // * The pattern ``(?i)/xxx/`` paired with a substitution string of ``/yyy/``
+  //   would do a case-insensitive match and transform path ``/aaa/XxX/bbb`` to
+  //   ``/aaa/yyy/bbb``.
+  RegexRewrite regex_rewrite = 32;
```

Comment:
I'm OK with this API as I agree it's the most flexible, and in general I don't think we are trying to go max perf with regex routing since there are known perf issues which come from using it. @lizan any objections here? Assuming we go with this:
1) I agree with @zuercher that we should promtoe this field and prefix_rewrite to be part of a oneof.
2) Can we clarify in the docs here that prefix_rewrite takes precedence, etc. (modify both the docs for prefix_rewrite as well as this doc).

Reply:
I've updated the docs for prefix_rewrite and regex_rewrite to say that only one should be specified, and that prefix_rewrite takes precedence.

I didn't make the request annotation change because I'm not exactly sure where it goes.

Reply:
> I didn't make the request annotation change because I'm not exactly sure where it goes.

Actually I think we can't do it right now as we need to do it on v3 and have it show up for v4, which the tooling does not support yet. Can you open a tech debt issue on this so we can track it once this PR lands? Thank you. cc @lizan

Reply:
By my reading of api/STYLE.md, I think the idea is that we mark both `prefix_rewrite` and `regex_rewrite` with 

```
[(udpa.annotations.field_migrate).oneof_promotion = "path_rewrite"]
```

That indicates that we'll promote those fields to a oneof block in v4. Also, rather than having `regex_rewrite` take precedence over `prefix_rewrite`, throw an EnvoyException if both are set to give v2 and v3 oneof semantics for these fields in the meantime.

Reply:
And I think that the annotation only goes on the v3 version of the proto, not the v2.

Reply:
I've updated the code to raise an exception if both `prefix_rewrite` and `regex_rewrite` are specified, and updated the documentation to match.  I also added the `oneof_promotion` annotation as described.